src/js/angular/resources/controllers.js

```diff
@@ -65,211 +66,215 @@ resourcesCtrl.controller('ResourcesCtrl', ['$scope', '$timeout', 'MonitoringRest
-                        $scope.firstLoad = false;
-
-                        const timer = $timeout(function () {
-                            $scope.loader = false;
-                        }, 500);
 
-                        $scope.$on('$destroy', function () {
-                            $timeout.cancel(timer);
-                        });
+        // TODO: Organize the following somehow
+        const getData = (monitor) => {
+            if (hasMonitorError(monitor.error)) {
+                if (monitor.error.retries === MAX_RETRIES) {
+                    if (monitor.poll) {
+                        $timeout.cancel(monitor.poll);
                     }
+                    return;
                 }
-                structuresMonitorPoll = $timeout(getStructuresMonitorData, POLLING_INTERVAL);
-            }).catch(function (data) {
-                $scope.error = getError(data);
-                $scope.loader = false;
-            });
+                monitor.error.retries++;
+            }
+            return monitor.fetchFn()
+                .then(() => {
+                    clearError(monitor.error);
+                })
+                .catch((error) => {
+                    setError(monitor.error, error);
+                })
+                .finally(() => {
+                    if (monitor.poll) {
+                        $timeout.cancel(monitor.poll);
+                    }
+                    monitor.poll = $timeout(() => getData(monitor), POLLING_INTERVAL);
+                });
         };
-        let clusterMonitorPoll;
-        const getClusterMonitorData = function () {
-            if ($scope.error) {
-                return;
+        const processResponse = (response, dataSetter) => {
+            const data = response.data;
+            if (data) {
+                setChartData(data, dataSetter);
+
+                if ($scope.firstLoad) {
+                    $scope.firstLoad = false;
+
+                    const timer = $timeout(function() {
+                        $scope.loader = false;
+                    }, 500);
+
+                    $scope.$on('$destroy', function() {
+                        $timeout.cancel(timer);
+                    });
+                }
             }
-            MonitoringRestService.monitorCluster().then(function (response) {
-                const data = response.data;
-                if (data) {
-                    const timestamp = new Date();
+        };
+        const setChartData = (data, dataSetter) => {
+            const timestamp = new Date();
+            dataSetter(timestamp, data);
+        };
+        const setError = (errorHolder, error) => {
+            errorHolder.hasError = !!error;
+            if (errorHolder.hasError) {
+                errorHolder.message = getError(error);
+            }
+            $scope.error = Object.values($scope.monitors).some((monitor) => monitor.error.hasError);
+        };
+        const clearError = (errorHolder) => {
```

Comment:
Maybe this should be `updateError` because you may not clear the error. Right?

Reply:
It should be `clearMonitorError` as what it does is to clear the monitor error. The check at the end is just an update to the `error` flag